api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,758 @@
+	"github.com/quintilesims/layer0/common/errors"
+	"github.com/quintilesims/layer0/common/models"
+)
+
+var defaultPorts = []int{
+	22,
+	2376,
+	2375,
+	51678,
+	51679,
+}
+
+// https://aws.amazon.com/ec2/instance-types/
+var InstanceSizes = map[string]bytesize.Bytesize{
+	"t2.nano":     0.5 * bytesize.GiB,
+	"t2.micro":    1 * bytesize.GiB,
+	"t2.small":    2 * bytesize.GiB,
+	"t2.medium":   4 * bytesize.GiB,
+	"t2.large":    8 * bytesize.GiB,
+	"m4.large":    8 * bytesize.GiB,
+	"m4.xlarge":   16 * bytesize.GiB,
+	"m4.2xlarge":  32 * bytesize.GiB,
+	"m4.4xlarge":  64 * bytesize.GiB,
+	"m4.10xlarge": 160 * bytesize.GiB,
+	"m3.medium":   3.75 * bytesize.GiB,
+	"m3.large":    7.5 * bytesize.GiB,
+	"m3.xlarge":   15 * bytesize.GiB,
+	"m3.2xlarge":  30 * bytesize.GiB,
+	"c4.large":    3.75 * bytesize.GiB,
+	"c4.xlarge":   7.5 * bytesize.GiB,
+	"c4.2xlarge":  15 * bytesize.GiB,
+	"c4.4xlarge":  30 * bytesize.GiB,
+	"c4.8xlarge":  60 * bytesize.GiB,
+	"c3.large":    3.75 * bytesize.GiB,
+	"c3.xlarge":   7.5 * bytesize.GiB,
+	"c3.2xlarge":  15 * bytesize.GiB,
+	"c3.4xlarge":  30 * bytesize.GiB,
+	"c3.8xlarge":  60 * bytesize.GiB,
+	"g2.2xlarge":  15 * bytesize.GiB,
+	"g2.8xlarge":  60 * bytesize.GiB,
+	"x1.32xlarge": 1952 * bytesize.GiB,
+	"r3.large":    15.25 * bytesize.GiB,
+	"r3.xlarge":   30.5 * bytesize.GiB,
+	"r3.2xlarge":  61 * bytesize.GiB,
+	"r3.4xlarge":  122 * bytesize.GiB,
+	"r3.8xlarge":  244 * bytesize.GiB,
+	"i3.large":    15.25 * bytesize.GiB,
+	"i3.xlarge":   30.5 * bytesize.GiB,
+	"i3.2xlarge":  61 * bytesize.GiB,
+	"i3.4xlarge":  122 * bytesize.GiB,
+	"i3.8xlarge":  244 * bytesize.GiB,
+	"d2.xlarge":   30.5 * bytesize.GiB,
+	"d2.2xlarge":  61 * bytesize.GiB,
+	"d2.4xlarge":  122 * bytesize.GiB,
+	"d2.8xlarge":  244 * bytesize.GiB,
+}
 
 type EnvironmentScaler struct {
+	Client              *awsc.Client
+	EnvironmentProvider provider.EnvironmentProvider
+	ServiceProvider     provider.ServiceProvider
+	TaskProvider        provider.TaskProvider
+	JobStore            job.Store
+	Config              config.APIConfig
+	deployCache         map[string][]ResourceConsumer
 }
 
-func NewEnvironmentScaler() *EnvironmentScaler {
-	return &EnvironmentScaler{}
+const (
+	Memory = iota
+	CPU
+)
+
+func NewEnvironmentScaler(a *awsc.Client, e provider.EnvironmentProvider, s provider.ServiceProvider, t provider.TaskProvider, j job.Store, c config.APIConfig) *EnvironmentScaler {
+	return &EnvironmentScaler{
+		Client:              a,
+		EnvironmentProvider: e,
+		ServiceProvider:     s,
+		TaskProvider:        t,
+		JobStore:            j,
+		Config:              c,
+		// deployCache: map[string][]*Resource,
+	}
 }
 
 func (e *EnvironmentScaler) Scale(environmentID string) error {
-	return fmt.Errorf("EnvironmentScaler not implemented")
+	clusterName := addLayer0Prefix(e.Config.Instance(), environmentID)
+
+	resourceProviders, err := e.getResourceProviders(clusterName)
+	if err != nil {
+		return err
+	}
+
+	humanReadableProviders := ""
```

Comment:
@zpatrick With `[]*ResourceProvider`, all `%#v` gave me was a list of memory addresses, which is very opaque and unhelpful. Do you know of another hidden trick of `%#v` that I don't? Otherwise, it seems like `String()` is the way to go.

Reply:
Implemented `String()` method on `*ResourceProvider`. ✅

Reply:
hmm, not that I'm aware of, but probably worth investigating for future use-cases.

Reply:
FWIW, consensus seems to be that `String()` is the right thing to do here.